File: Easy/Isomorphic Strings/isomorphic-strings.cpp
//{ Driver Code Starts
// C++ program to check if two strings are isomorphic


#include<bits/stdc++.h>


using namespace std;
#define MAX_CHARS 256

// } Driver Code Ends
class Solution
{
    public:
    //Function to check if two strings are isomorphic.
    bool areIsomorphic(string str1, string str2)
    {
        
        // Your code here
        unordered_map<char, int> mp1;
        unordered_map<char, int> mp2;
        
        int count = 0;
        
        for(int i = 0; i < str1.length(); i++)
        {
            if(mp1.find(str1[i]) == mp1.end())
            {
            mp1[str1[i]] = count;
            count++; 
            }
            
        }
        
        count = 0;
        
        for(int i = 0; i < str2.length(); i++)
        {
            if(mp2.find(str2[i]) == mp2.end())
            {
            mp2[str2[i]] = count;
            count++; 
            }
        }
        
        // int count = 0;
        
        // for(auto itr = mp1.begin(); itr != mp1.end(); itr++)
        // {
        //     itr->second = count;
        //     count++;
        // }
        
        // count = 0;
        // for(auto itr = mp2.begin(); itr != mp2.end(); itr++)
        // {
        //     itr->second = count;
        //     count++;
        // }
        
        vector<int> vec1;
        vector<int> vec2;
        
        for(int i = 0; i < str1.length(); i++)
        {
           vec1.emplace_back(mp1[str1[i]]); 
        }
        
        for(int i = 0; i < str2.length(); i++)
        {
            vec2.emplace_back(mp2[str2[i]]);
        }
        
        if(vec1 == vec2)
        {
            return true;
        }
        
        return false;
        
        
    }
};

//{ Driver Code Starts.

// Driver program
int main()
{
    int t;
    cin>>t;
    string s1,s2;
    while (t--) {
        cin>>s1;
        cin>>s2;
        Solution obj;
        cout<<obj.areIsomorphic(s1,s2)<<endl;
    }
    
    return 0;
}
// } Driver Code Ends

File: Easy/Isomorphic Strings/isomorphic-strings.java
//{ Driver Code Starts
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.*;


public class Driver {
    
    public static void main(String[] args)throws IOException {
       
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine());
        while(tc-->0)
        {
            String s1 = br.readLine();
            String s2 = br.readLine();
            
            Solution obj = new Solution();
            
            boolean a = obj.areIsomorphic(s1,s2);
            if(a)
            System.out.println(1);
            else
            System.out.println(0);
        }
    }
    
}

// } Driver Code Ends


class Solution
{
    //Function to check if two strings are isomorphic.
    public static boolean areIsomorphic(String str1,String str2)
    {
        // Your code here
         if(str1.length() != str2.length())
        {
            return false;
        }
        
        HashMap<Character, Character> hm = new HashMap<>();
        int arr[] = new int [26];
        
        for(int i = 0; i < str1.length(); i++)
        {
            char a = str1.charAt(i);
            char b = str2.charAt(i);
            
            if(hm.containsKey(a))
            {
                if(!hm.get(a).equals(b))
                {
                    return false;
                }
            }
            else
            {
                if(arr[b - 'a'] == 1)
                {
                    return false;
                }
                arr[b - 'a'] = 1;
                hm.put(a,b);
            }
        }
        
        return true;
        
    }
}